libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,607 @@
+    // @User: Advanced
+    AP_GROUPINFO("BARO_ACC", 5, AP_OpenDroneID, _baro_accuracy, -1.0),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_OpenDroneID::AP_OpenDroneID()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("OpenDroneID must be singleton");
+    }
+#endif
+    _singleton = this;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_OpenDroneID::init()
+{
+    if (_enable == 0) {
+        return;
+    }
+
+    _chan = mavlink_channel_t(gcs().get_channel_from_port_number(_mav_port));
+    _mavlink_dynamic_period_ms = 1000; // send dynamic messages at 1 Hz frequency i.e. every 1 second
+    _mavlink_static_period_ms = 10000; // send dynamic messages at 0.1 Hz frequency i.e. every 10 seconds
+}
+
+// Perform the pre-arm checks and prevent arming if they are not satisifed
+// Except in the case of an in-flight reboot
+bool AP_OpenDroneID::pre_arm_check(char* failmsg, uint8_t failmsg_len)
+{
+    WITH_SEMAPHORE(_sem);
+
+    if (!option_set(Options::EnforceArming)) {
+        return true;
+    }
+
+    if (pkt_basic_id.id_type == MAV_ODID_ID_TYPE_NONE) {
+        strncpy(failmsg, "UA_TYPE required in BasicID", failmsg_len);
+        return false;
+    }
+
+    if (pkt_system.operator_latitude == 0 && pkt_system.operator_longitude == 0) {
+        strncpy(failmsg, "operator location must be set", failmsg_len);
+        return false;
+    }
+
+    const uint32_t max_age_ms = 3000;
+    const uint32_t now_ms = AP_HAL::millis();
+
+    if (last_arm_status_ms == 0 || now_ms - last_arm_status_ms > max_age_ms) {
+        strncpy(failmsg, "ARM_STATUS not available", failmsg_len);
+        return false;
+    }
+
+    if (last_system_ms == 0 || now_ms - last_system_ms > max_age_ms) {
+        strncpy(failmsg, "SYSTEM not available", failmsg_len);
+        return false;
+    }
+    
+    if (arm_status.status != MAV_ODID_GOOD_TO_ARM) {
+        strncpy(failmsg, arm_status.error, failmsg_len);
+        return false;
+    }
+    
+    return true;
+}
+
+void AP_OpenDroneID::update()
+{
+    if (_enable == 0) {
+        return;
+    }
+    const uint32_t now = AP_HAL::millis();
+
+    const bool armed = hal.util->get_soft_armed();
+    if (armed && !_was_armed) {
+        // use arm location as takeoff location
+        AP::ahrs().get_location(_takeoff_location);
+    }
+    _was_armed = armed;
+
+    if (now - _last_send_dynamic_messages_ms >= _mavlink_dynamic_period_ms) {
+        _last_send_dynamic_messages_ms = now;
+        send_dynamic_out();
+    }
+
+    send_static_out();
+}
+
+void AP_OpenDroneID::send_dynamic_out()
+{
+    send_location_message();
+}
+
+void AP_OpenDroneID::send_static_out()
+{
+    const uint32_t now_ms = AP_HAL::millis();
+
+    // we need to notify user if we lost the transmitter
+    if (now_ms - last_arm_status_ms > 5000) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "ODID: lost transmitter");
+    }
+
+    const uint32_t msg_spacing_ms = 5000 / 4;
```

Comment:
Why is the interval hard-coded and not using _mavlink_static_period_ms?

Reply:
fixed